Modules/Sources/Yosemite/Tools/POS/POSCatalogFullSyncService.swift

```diff
@@ -102,4 +114,61 @@ private extension POSCatalogFullSyncService {
         return POSCatalog(products: products, variations: variations, syncDate: syncStartDate)
     }
 
+    func loadCatalogFromCatalogAPI(for siteID: Int64, syncRemote: POSCatalogSyncRemoteProtocol, regenerateCatalog: Bool) async throws -> POSCatalog {
+        let downloadStartTime = CFAbsoluteTimeGetCurrent()
+        let catalog = try await downloadCatalog(for: siteID, syncRemote: syncRemote, regenerateCatalog: regenerateCatalog)
+        let downloadTime = CFAbsoluteTimeGetCurrent() - downloadStartTime
+        DDLogInfo("🟣 Catalog download completed - Time: \(String(format: "%.2f", downloadTime))s")
+
+        return .init(products: catalog.products, variations: catalog.variations, syncDate: .init())
+    }
+}
+
+private extension POSCatalogFullSyncService {
+    func downloadCatalog(for siteID: Int64, syncRemote: POSCatalogSyncRemoteProtocol, regenerateCatalog: Bool) async throws -> POSCatalogResponse {
+        DDLogInfo("🟣 Starting catalog request...")
+
+        // 1. Requests catalog until download URL is available.
+        let response = try await syncRemote.requestCatalogGeneration(for: siteID, forceGeneration: regenerateCatalog)
+        let downloadURL: String?
+        if let url = response.downloadURL {
+            downloadURL = url
+        } else {
+            // 2. Polls for completion until download URL is available.
+            downloadURL = try await pollForCatalogCompletion(siteID: siteID, syncRemote: syncRemote)
+        }
+
+        // 3. Downloads catalog using the provided URL.
+        guard let downloadURL else {
+            throw POSCatalogSyncError.invalidData
+        }
+        DDLogInfo("🟣 Catalog ready for download: \(downloadURL)")
+        return try await syncRemote.downloadCatalog(for: siteID, downloadURL: downloadURL)
+    }
+
+    func pollForCatalogCompletion(siteID: Int64, syncRemote: POSCatalogSyncRemoteProtocol) async throws -> String {
+        // Each attempt is made 1 second after the last one completes.
+        let maxAttempts = 1000
+        var attempts = 0
+
+        while attempts < maxAttempts {
+            let response = try await syncRemote.requestCatalogGeneration(for: siteID, forceGeneration: false)
+
+            switch response.status {
+            case .complete:
+                guard let downloadURL = response.downloadURL else {
+                    throw POSCatalogSyncError.invalidData
+                }
+                return downloadURL
+            case .pending, .processing:
+                DDLogInfo("🟣 Catalog request \(response.status)... (attempt \(attempts + 1)/\(maxAttempts))")
```

Comment:
This will flood the logs, for large catalogs... 

Currently we log with `.verbose` even in prod. Perhaps we should change that to `.info` in AppDelegate, and then update these to be Debug logs?

Reply:
Good catch! As an alternative to making global changes for the whole app, I reduced the logging noise by only logging every 10th polling attempt in https://github.com/woocommerce/woocommerce-ios/pull/16301/commits/e1990ed5dcdcd7d823e3bdf8c418c636086cf774. This keeps the logs useful for debugging (every ~10 seconds with 1s polling or ~50 seconds with 5s backoff polling later) without flooding them for long-running catalog generations. WDYT?

Reply:
That's fine, yeah... but I do think we should change this separately anyway.

IMO, the right way to do this would be to make this a `.debug` or even `.verbose` level log. Then we keep debug builds log output set to `verbose` and set production logging to `info`. We'd have to check there aren't other places where we're using debug log levels that we want for debugging customer issues, so it makes sense not to do that in this PR.